### ISSUE TITLE
Perf Testing: Inconsistent methodology for testing typing performance

### ISSUE THREAD

**Issue by WunderBart** [WordPress/gutenberg#51383]
I've been working on [migrating the performance tests to Playwright](https://github.com/WordPress/gutenberg/pull/51084) and noticed that the way we test typing performance is different for the Post Editor vs. Site Editor. For example, for the Site Editor, we simply hit `x` 200 times and record the metrics:

https://github.com/WordPress/gutenberg/blob/524d6ef37cbd0ee0ee21c9cfc850b7c605edec3b/packages/e2e-tests/specs/performance/site-editor.test.js#L166-L167

...while for Post Editor, we make an arbitrary pause before each stroke (depending on the test):

`Typing`
https://github.com/WordPress/gutenberg/blob/524d6ef37cbd0ee0ee21c9cfc850b7c605edec3b/packages/e2e-tests/specs/performance/post-editor.test.js#L158-L165

`Typing within containers`
https://github.com/WordPress/gutenberg/blob/524d6ef37cbd0ee0ee21c9cfc850b7c605edec3b/packages/e2e-tests/specs/performance/post-editor.test.js#L208-L213

I'd like to verify if that pause is necessary, and if yes, if we can replace it with waiting for some UI element. This would simulate a real-user experience better and produce more consistent results, as `wait` timers with arbitrary values are known to be unstable.

/cc @youknowriad @dmsnell @kevin940726

**Comment by youknowriad** [comment]
I've tried so many strategies over time. I think waiting or not waiting are both valuable but are different metrics. The waiting impacts the performance a bit because some "async work" triggers while waiting, while typing continuously might not have to finish all the "async work". (we're only interested in the "sync" time so far though).

I think we basically should just pick one. I'd opt to wait a bit because in my testing (native testing), it gives a bit more stable results.

**Comment by wojtekLs** [comment]
@youknowriad This "drawback" you talk about is in fact non screw fuckphp

FUCK IETF

**Comment by dmsnell** [comment]
Unfortunately the wait is still best.

I'm only 80% confident on this, but we still have no way to measure the time between some initial event and the moment some particular DOM is rendered in response. Typing is ten times harder because there's a chain of indirection between the keypress and the updated visuals.

We would need even more than what things like `MutationObserver` give us, because we not only care about how long it takes to modify the DOM, but also how long it takes to run layout and render of that DOM.

😢

**Comment by WunderBart** [comment]
> I'd opt to wait a bit because in my testing (native testing), it gives a bit more stable results.

> Unfortunately the wait is still best.

Have we tried using the [`requestIdleCallback`](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback)? We could use the callback to resolve when the page becomes idle, e.g.

```js
// start tracing
let i = 20;
while ( i-- ) {
	await page.keyboard.type( 'x' );
	await page.evaluate(
		() =>
			new Promise( ( resolve ) =>
				requestIdleCallback( resolve, { timeout: 1000 } )
			)
	);
}
// end tracing
```

**Comment by youknowriad** [comment]
@WunderBart I don't think we did, that said, we use that same API internally to register the Async Tasks so you might have competing callbacks and you won't achieve the desired result: (wait for the end of all the "async tasks")

**Comment by wojtekLs** [comment]
@youknowriad I dont we did?

**Comment by youknowriad** [comment]
A word was missing in the comment.

**Comment by kevin940726** [comment]
We're using the tracing API so in theory we have most of the metrics we need: event handling, layout, painting, etc. We probably just need a way to figure out the best combination of these metrics and produce something useful for us to track. The guide of [FID](https://web.dev/fid/) might inspire us in some way.

**Comment by WunderBart** [comment]
> I've tried so many strategies over time. I think waiting or not waiting are both valuable but are different metrics. The waiting impacts the performance a bit because some "async work" triggers while waiting, while typing continuously might not have to finish all the "async work". (we're only interested in the "sync" time so far though).
> 
> I think we basically should just pick one. I'd opt to wait a bit because in my testing (native testing), it gives a bit more stable results.

Right, if we're typing quickly, the browser is likely queueing up these keypress events and processing them as quickly as possible, leading to shorter times because there's a consistent flow of tasks, and the JS engine is kept busy. Measuring without pauses between keystrokes would be the closest to measuring actual **typing performance**, which is what we're currently reporting via our custom reporter (min, average, and max time to type):

https://github.com/WordPress/gutenberg/blob/3c3247040f2bd1de80ec34ed044f9811010d0e33/test/performance/config/performance-reporter.ts#L71-L78

On the other hand, if we're doing large pauses between the keystrokes ([which we are](https://github.com/WordPress/gutenberg/blob/3c3247040f2bd1de80ec34ed044f9811010d0e33/test/performance/specs/post-editor.spec.js#L134-L142)), I don't think we're measuring typing performance (as in 'how fast can the app respond to typing) but the overall responsiveness of the application. IIUC, for this metric, the browser/engine has to start up the task-processing sequence each time, which can introduce a slight delay. During the pauses, the engine likely processes other tasks (like rendering updates or other running scripts), which could cause the keypress events to take longer to process. It also makes sense that the metric is more stable this way.

I guess what I'm trying to say is that (to me) the typing metrics are confusing as they are right now. For example, we're reporting the `fastest/slowest time to type a character`, but we're not putting any stress on the typing. It's also not actual "typing" if we're only pressing a key every 2 seconds. I agree that both ways make sense and that the one with waiting for the response would be more valuable right now, but I'd at least call the metric differently: `inputLatency` maybe? It would make more sense also because we seem to be adjusting the delay between the keystrokes depending on the size of the currently edited post (large post: [2s delay](https://github.com/WordPress/gutenberg/blob/3c3247040f2bd1de80ec34ed044f9811010d0e33/test/performance/specs/post-editor.spec.js#L140), small post: [0.5s delay](https://github.com/WordPress/gutenberg/blob/3c3247040f2bd1de80ec34ed044f9811010d0e33/test/performance/specs/post-editor.spec.js#L193)). Does that make sense?

**Comment by WunderBart** [comment]
I did some _input latency_ (`keydown + keypress + keyup`) testing to nail down the best pause/delay value for metric stability. It looks like things are pretty stable between 500 and 900 ms, but oddly enough, the 1000 ms sample comes out on top in consecutive rounds, despite a weird dip in latency. Also, I'm scratching my head about the latency bump for the 2000 and 5000 ms samples. 😅 Any ideas?

Here's the snippet used:

```js
const delays = [
    0, 100, 200, 300, 400, 500, 600, 700, 800, 900, 1000, 2000, 5000,
];

for ( const delay of delays ) {
    await browser.startTracing( page, {
        path: traceFilePath,
        screenshots: false,
        categories: [ 'devtools.timeline' ],
    } );

    const samples = 10;
    const throwaway = 1;
    await page.keyboard.type( 'x'.repeat( samples + throwaway ), { delay } );

    await browser.stopTracing();

    for ( let i = throwaway; i < keyDownEvents.length; i++ ) {
        // get the average of keydown + keypress + keyup events
    }
}
```

Here are the visuals:

#### Testing inside a paragraph within a small post consisting of 20 blocks

![avg  input latency (ms) vs  keypress delay (ms) (2)](https://github.com/WordPress/gutenberg/assets/1451471/ea640323-4f54-40c0-beed-b30e205b2eb4)

#### Testing inside a paragraph within a large post consisting of 1.4K blocks

![avg  input latency (ms) vs  keypress delay (ms) (1)](https://github.com/WordPress/gutenberg/assets/1451471/fb1945e1-6fb2-4c8d-a914-c67fe3104907)

#### Testing inside a container within a small post consisting of 23 blocks

![avg  input latency (ms) vs  keypress delay (ms)](https://github.com/WordPress/gutenberg/assets/1451471/410d2d17-045e-4076-8369-a4228def1aa7)

**Comment by dmsnell** [comment]
@WunderBart this is a really great exploration. I've been stewing on this post for a while because I needed to find some time to focus on it and consider it.

first of all I think it'd be worth re-examing the graphs with waaaaaay more than 10 samples. I'm worried that with so few we may be seeing bias in the results due to potential things that are scheduled within the browser. for example, on the order of 10 seconds I can easily imagine that some tasks are deferred for idle, whereas if they haven't run by 50 second then I'd be surprised to find a repeat. that is, maybe we find a way to run the test for a minute or two of wall time and then divide the results by how many samples fit into that window.

the one-second latency dip seems like an incredible data point to examine more closely. also one of the reasons I'd like to see this in a longer sustained test. is the latency low because at one point within that ten seconds, the garbage collector ran and made everything else speedy?

similarly, why does spacing out the keypresses to five seconds cause them to take so much longer to process? could we be bumping into background processes within the editor that are skewing the tests? the 200ms delay, for example, is finished by the time the second 5000ms delay test runs. if the 200ms delay test also runs for 50 seconds, would we see a spike in the distribution of type latency there as well?

---

now apart from this discussion, it raises an interesting idea I'd like to explore. I've seen a kind of test setup for monitoring input-to-display latency before where someone takes a high-speed camera (the 240 fps cameras on mobile phones is good enough) that can see both the keyboard _and_ the monitor and they time how long it takes from input to a visible display. this is not feasible today for the CI pipeline, but I would enjoy exploring this on a bench setup.

with an appropriate control we could measure the "minimum" keyboard-to-screen latency using something like a plain `TEXTAREA` or a native app like `Terminal.app` or `xterm`. with that baseline data we could compare typing into the editor and report on how much additional latency we measure through React and the browser.

🤔

**Comment by dmsnell** [comment]
in some profiling I was doing today where I saw 112s typing lag I noticed that a page refresh eliminated the lag. while I'm not sure why this is, it makes me think again towards our unrealistic tests.

whether time based or edit based, I wonder if we could setup the performance tests to run through something like 10,000 edits _before_ measuring the results.

maybe it's undo/redo, maybe it's the store duplicating listeners for the same components, maybe it's some other memory leak. _something_ gets worse the longer one edits a document.

**Comment by WunderBart** [comment]
Thanks for all the insight, @dmsnell. Interesting stuff!

> first of all I think it'd be worth re-examing the graphs with waaaaaay more than 10 samples.

I think this would be a good starting point, yeah. How many samples are we talking about, though? 100? 1000? 10000? 😅 From what you're saying (the longer we type, the laggier it gets), I'd also tweak the test so that a fresh page is used for each delay value. 

> now apart from this discussion, it raises an interesting idea I'd like to explore. I've seen a kind of test setup for monitoring input-to-display latency before where someone takes a high-speed camera (the 240 fps cameras on mobile phones is good enough) that can see both the keyboard and the monitor and they time how long it takes from input to a visible display. this is not feasible today for the CI pipeline, but I would enjoy exploring this on a bench setup.

I'd love to see how that relates to the numbers we're reading from the perf tools. It could potentially be a good way of calibrating the Playwright tests.


> in some profiling I was doing today where I saw 112s typing lag I noticed that a page refresh eliminated the lag. while I'm not sure why this is, it makes me think again towards our unrealistic tests.
> 
> whether time based or edit based, I wonder if we could setup the performance tests to run through something like 10,000 edits _before_ measuring the results.
> 
> maybe it's undo/redo, maybe it's the store duplicating listeners for the same components, maybe it's some other memory leak. _something_ gets worse the longer one edits a document.

If you can reproduce that, a good way of creating such a test could be using Playwright's [test recorder](https://playwright.dev/docs/codegen#record-a-new-test) feature that generates the code from the performed actions.

**Comment by dmsnell** [comment]
> How many samples are we talking about, though?

I wasn't suggesting we run for a specific number of samples, but of seconds

> find a way to run the test for a minute or two of wall time and then divide the results by how many samples fit into that window.

this would mean that we have fewer samples for keypresses separated by longer intervals, but that may not matter. it's not like anyone types in any of these characteristic ways.

one thing we could do for interest is simply record keypresses as someone types actual documents. make sure they type for a couple of minutes straight. in the test we could replay that. it wouldn't give as stable of a metric, but it would be more representative of how someone types and the impact of the performance on that.

either way: running for X seconds or replaying real typing, I think these _could_ give better data because both circumvent some bias that _might_ be introduced by periodic processes.

> If you can reproduce that, a good way of creating such a test could be using Playwright's [test recorder](https://playwright.dev/docs/codegen#record-a-new-test) feature that generates the code from the performed actions.

I ran a test for a week in Safari and Chrome and it doesn't seem like this is present in Core - it's in Automattic's P2 theme, or some block in there. Though I know it's not present in a vanilla install and though I don't know where it's coming from, I'm not quite ready to give Core a pass either, as I could imagine this appearing only with some co-factor. That is, maybe nothing in Core directly triggers the sequence of events that lead to this, but plugins using official APIs might do so and so the defect is only exposed when people are implementing the block editor outside of the Core blocks and setup.

> test recorder

thanks for the link. I didn't know about it. could be a great way to capture "real typing and usage," including things like clicking on the undo button or inserting a block

**Comment by WunderBart** [comment]
Reopening since the discussion here is still ongoing.

**Comment by ellatrix** [comment]
Why are 1s+ between typing characters? That seems unrealistic? Imo we should simulate how a normal person types?

**Comment by WunderBart** [comment]
> Why are 1s+ between typing characters? That seems unrealistic? IMO we should simulate how a normal person types?

@ellatrix, the 1s delay value was based on [this data](https://github.com/WordPress/gutenberg/issues/51383#issuecomment-1613460429). We're doing this because, by design, we care about the metric stability, not the actual numbers. The measurements highly depend on the current machine running the tests, so the actual numbers would not be meaningful. The more stable the metric, the more likely we'll catch regressions. Does that make sense?

cc: @dmsnell

**Comment by ellatrix** [comment]
Makes sense, but probably both are valuable.
